scalanet/discovery/src/io/iohk/scalanet/discovery/ethereum/v4/DiscoveryService.scala

```diff
@@ -1,9 +1,798 @@
+          state.withLastPongTimestamp(peer, now)
+        }
+      }
+
+    /** Update the bonding state of the peer with the result,
+      * notifying all potentially waiting bonding processes about the outcome.
+      */
+    protected[v4] def completePong(peer: Peer[A], responded: Boolean): Task[Unit] =
+      stateRef
+        .modify { state =>
+          val maybePongReceived = state.bondingResultsMap.get(peer).map(_.pongReceived)
+          state.clearBondingResults(peer) -> maybePongReceived
+        }
+        .flatMap { maybePongReceived =>
+          maybePongReceived.fold(Task.unit)(_.complete(responded))
+        }
+
+    /** Allow the remote peer to ping us during bonding, so that we can have a more
+      * fungible expectation that if we send a message they will consider us bonded and
+      * not ignore it.
+      *
+      * The deferred should be completed by the ping handler.
+      */
+    protected[v4] def awaitPing(peer: Peer[A]): Task[Unit] =
+      stateRef.get
+        .map { state =>
+          state.bondingResultsMap.get(peer).map(_.pingReceived)
+        }
+        .flatMap { maybePingReceived =>
+          maybePingReceived.fold(Task.unit)(_.get.timeoutTo(config.requestTimeout, Task.unit))
+        }
+
+    /** Complete any deferred we set up during a bonding process expecting a ping to arrive. */
+    protected[v4] def completePing(peer: Peer[A]): Task[Unit] =
+      stateRef.get
+        .map { state =>
+          state.bondingResultsMap.get(peer).map(_.pingReceived)
+        }
+        .flatMap { maybePingReceived =>
+          maybePingReceived.fold(Task.unit)(_.complete(()).attempt.void)
+        }
+
+    /** Fetch the remote ENR if we don't already have it or if
+      * the sequence number we have is less than what we got just now.
+      *
+      * The execution might be delayed in case we are expecting the other side
+      * to receive our Pong first, lest they think we are unbonded.
+      * Passing on the variable so the Deferred is entered into the state.
+      *  */
+    protected[v4] def maybeFetchEnr(
+        peer: Peer[A],
+        maybeRemoteEnrSeq: Option[ENRSeq],
+        delay: Boolean = false
+    ): Task[Unit] =
+      for {
+        state <- stateRef.get
+        maybeEnrAndNode = (state.enrMap.get(peer.id), state.nodeMap.get(peer.id))
+        needsFetching = maybeEnrAndNode match {
+          case _ if state.isSelf(peer) =>
+            false
+          case (None, _) =>
+            true
+          case (Some(enr), _) if maybeRemoteEnrSeq.getOrElse(enr.content.seq) > enr.content.seq =>
+            true
+          case (_, Some(node)) if toAddress(node.address) != peer.address =>
+            true
+          case _ =>
+            false
+        }
+        _ <- fetchEnr(peer, delay).whenA(needsFetching)
+      } yield ()
+
+    /** Fetch a fresh ENR from the peer and store it.
+      *
+      * Use delay=true if there's a high chance that the other side is still bonding with us.
```

Comment:
It happened in your comprehensive integration tests. 
When Alice wants to talk to Bob, she has to bond with him first, and then Bob has to bond with Alice as well.
There's a sequence of events that happened all the time:

1. Alice wants to bond with Bob: Alice sends Ping to Bob
2. Bob receives the Alice's Ping and initiates an asynchronous bond as well: Bob sends a Ping to Alice.
3. Bob responds to Alice's Ping: Bob sends a Pong to Alice (may go before the Ping).
4. Alice receives Bob's Pong. She considers them bonded now and initiates an async ENR fetch: Alice sends ENRRequest to Bob.
5. Bob receives Alice's ENRRequest but he hasn't got a Pong yet so ignores it, they aren't yet bonded as far as he's concerned.
6. Alice receives Bob's Ping: Alice sends Pong to Bob. Maybe she also initiates an async ENR fetch.
7. Bob receives Alice's Pong. Now he considers her bonded. This may happen after the ENRRequest arrives.

Due to the way the `DiscoveryRPC` works we should be careful of sending messages from the handlers themselves: they may arrive earlier than the response given by the handler. For example in the `ping` handler we can start a `fetchEnr` but that may result in a message before the result of the `ping` invocation, the `Pong`, reaches the peer. While this isn't ideal, since this is UDP, we wouldn't get any ordering guarantee anyway, so if we quickly do a `send(Pong); send(ENRRequest)` they still may end up arriving in a different order. Therefore I figured the simplest way to avoid the problem is to introduce a little delay in situations when we expect that the other party may be pinging us, waiting for our pong.

Another option would be to do retries. I just did this to make the integration tests pass.

Reply:
ok, so lets leave it as it is and we will check how it works in practice